ESPResSo 5.0-dev

jngrad opened the pull request:
Build the development version of ESPResSo against foss/2023b.

jngrad committed:
ESPResSo 5.0-dev

Neves-P commented:
bot: show_config

Neves-P commented:
bot: show_config

Neves-P commented:
bot: show_config

Neves-P commented:
bot: show_config

Neves-P commented:
bot: show_config

Neves-P commented:
bot: show_config

Neves-P commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P committed:
Use full commit SHA and update source download file name

Neves-P commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P committed:
Add missing

Neves-P commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P commented:
```
== processing EasyBuild easyconfig /eessi_bot_job/easyconfigs/ESPResSo-foss-2023b-software-commit.eb
== building and installing ESPResSo/-4b9ccc45713137eb16e5f19e5a313b5342080c75-foss-2023b...
  >> installation prefix: /cvmfs/dev.eessi.io/espresso/versions/2023.06/software/linux/x86_64/amd/zen2/software/ESPResSo/-4b9ccc45713137eb16e5f19e5a313b5342080c75-foss-2023b
== fetching files...
== Running pre-fetch hook...
  >> download succeeded: https://github.com/jngrad/espresso/archive/4b9ccc45713137eb16e5f19e5a313b5342080c75.tar.gz
  >> download failed: https://i10git.cs.fau.de/api/v4/projects/walberla%2Fwalberla/repository/archive?sha=59c9b8b1/walberla-59c9b8b1.tar.gz
  >> download failed: https://sources.easybuild.io/e/ESPResSo/walberla-59c9b8b1.tar.gz
== ... (took 1 secs)
== FAILED: Installation ended unsuccessfully (build directory: /cvmfs/dev.eessi.io/espresso/versions/2023.06/software/linux/x86_64/amd/zen2/software/ESPResSo/-4b9ccc45713137eb16e5f19e5a313b5342080c75-foss-2023b): build failed (first 300 chars): Couldn't find file walberla-59c9b8b1.tar.gz anywhere, and downloading it didn't work either... Paths attempted (in order): /eessi_bot_job/easyconfigs/e/ESPResSo/walberla-59c9b8b1.tar.gz, /eessi_bot_job/easyconfigs/ESPResSo/walberla-59c9b8b1.tar.gz, /eessi_bot_job/easyconfigs/walberla-59c9b8b1.tar.gz (took 1 secs)
== Results of the build can be found in the log file(s) /tmp/eb-mozetc_7/eb-po1shrc7/easybuild-ESPResSo--4b9ccc45713137eb16e5f19e5a313b5342080c75-20250612.160131.neTMC.log
Fetching files:  43% (3/7) ━━━━━━━╸           0:00:01 (walberla-59c9b8b1.tar.gz)ERROR: Build of /eessi_bot_job/easyconfigs/ESPResSo-foss-2023b-software-commit.eb failed (err: "build failed (first 300 chars): Couldn't find file walberla-59c9b8b1.tar.gz anywhere, and downloading it didn't work either... Paths attempted (in order): /eessi_bot_job/easyconfigs/e/ESPResSo/walberla-59c9b8b1.tar.gz, /eessi_bot_job/easyconfigs/ESPResSo/walberla-59c9b8b1.tar.gz, /eessi_bot_job/easyconfigs/walberla-59c9b8b1.tar.gz")
```
Fails with above error.

@jngrad, I edited the easyconfig and easystack to use the `%(software_commit)s` template, previous builds failed because they couldn't download the ESPReSso source from your fork (e.g.: https://github.com/multixscale/dev.eessi.io-espresso/pull/1#issuecomment-2967187821) and I hadn't noticed the template wasn't used. I did mess up the version name in the process, I'll fix that.

The correct ESPReSso sources are now grabbed, but as seen above the download of the other tarballs fails. It's probably just a matter of slightly adjusting the `sources` section of the easyconfig. If that's okay with you, I can do that and push directly here (as soon as I get home about to :bike: there now).

jngrad committed:
Adjust URLs

jngrad commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

jngrad commented:
@Neves-P feel free to adjust the sources as you see fit. I have changed the URL from my fork to upstream, since the commit in question is now available upstream. I also adjusted the waLBerla URL with a pound symbol to discard the automatically-added tarball filename.

Neves-P committed:
Add download_filename fields

Neves-P commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P committed:
Add missing comma

Neves-P commented:
I missed a comma... :confused: 

bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P committed:
Correct ESPReSso"s  GitHub repo URL

Neves-P commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P committed:
Fix espresso tarball name

Neves-P commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P committed:
Fix source tarball names and checksums

Neves-P commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P commented:
```== 2025-06-12 21:50:27,345 run.py:251 INFO Using /cvmfs/software.eessi.io/versions/2023.06/compat/linux/x86_64/bin/bash as shell for running cmd:  ./configure --prefix=/cvmfs/dev.eessi.io/espresso/versions/2023.06/software/linux/x86_64/amd/zen2/software/
ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75-foss-2023b  -DCMAKE_SKIP_RPATH=OFF -DESPRESSO_BUILD_TESTS=ON -D PYTHON_EXECUTABLE=$EBROOTPYTHON/bin/python 
== 2025-06-12 21:50:27,346 run.py:260 INFO running cmd:  ./configure --prefix=/cvmfs/dev.eessi.io/espresso/versions/2023.06/software/linux/x86_64/amd/zen2/software/ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75-foss-2023b  -DCMAKE_SKIP_RPATH=OFF -DESP
RESSO_BUILD_TESTS=ON -D PYTHON_EXECUTABLE=$EBROOTPYTHON/bin/python  
== 2025-06-12 21:50:27,443 build_log.py:171 ERROR EasyBuild crashed with an error (at easybuild/tools/build_log.py:111 in caller_info): cmd " ./configure --prefix=/cvmfs/dev.eessi.io/espresso/versions/2023.06/software/linux/x86_64/amd/zen2/software/ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75-foss-2023b  -DCMAKE_SKIP_RPATH=OFF -DESPRESSO_BUILD_TESTS=ON -D PYTHON_EXECUTABLE=$EBROOTPYTHON/bin/python " exited with exit code 127 and output:
/cvmfs/software.eessi.io/versions/2023.06/compat/linux/x86_64/bin/bash: line 1: ./configure: No such file or directory
 (at easybuild/tools/run.py:695 in parse_cmd_output)
```
We're getting further in the build, but now it's getting stuck in the configure step. I'm trying to work out what is causing it.

jngrad commented:
It looks like it's passing CMake flags to Autotools, and a few flags are also missing. The build should use `EB_ESPResSo`.

Neves-P commented:
> It looks like it's passing CMake flags to Autotools, and a few flags are also missing. The build should use `EB_ESPResSo`.

The build logs show the `EB_ESPResSo` easyblock being used, but indeed the wrong flags. I'm replicating the build interactively now to try and find out why.

Neves-P committed:
Consistent tarball namesb between steps

Neves-P committed:
Explicitly include this repo's easyblock

Neves-P commented:
You were totally right! From the log:
```
== 2025-06-12 21:50:14,629 easyblock.py:317 INFO This is easyblock EB_ESPResSo from module easybuild.easyblocks.espresso (/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/EasyBuild/4.9.4/lib/python3.11/site-packages/easybuild/easyblocks/e/espresso.py)
```
The easyblock in EB 4.9.4 was being pulled. I tried forcing it to use your new one now.

Neves-P commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P committed:
Proper path to easyblock

Neves-P commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P commented:
Failed because of https://github.com/EESSI/software-layer-scripts/pull/7

Let's try to work it out...

Neves-P committed:
Add EESSI version subdir in easystacks/ to account for EESSI/software-layer-scripts#7

Neves-P commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P committed:
Try pathname with wildcards

Neves-P commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P commented:
Edit: scratch that, that is not the actual error.

This is it:

```
[118/259] Linking CXX executable src/core/reaction_methods/tests/SingleReaction_test
FAILED: src/core/reaction_methods/tests/SingleReaction_test 
: && /cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/OpenMPI/4.1.6-GCC-13.2.0/bin/mpicxx --sysroot=/cvmfs/software.eessi.io/versions/2023.06/compat/linux/x86_64 -O2 -ftree-vectorize -march=native -fno-math-errno -fPIC -O
3 -DNDEBUG -L/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/HDF5/1.14.3-gompi-2023b/lib64 -L/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/HDF5/1.14.3-gompi-2023b/lib -L/cvmfs/software
.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/Pint/0.24-GCCcore-13.2.0/lib64 -L/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/Pint/0.24-GCCcore-13.2.0/lib -L/cvmfs/software.eessi.io/versions/2023.06
/software/linux/x86_64/amd/zen2/software/IPython/8.17.2-GCCcore-13.2.0/lib64 -L/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/IPython/8.17.2-GCCcore-13.2.0/lib -L/cvmfs/software.eessi.io/versions/2023.06/software/linux/
x86_64/amd/zen2/software/GSL/2.7-GCC-13.2.0/lib64 -L/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/GSL/2.7-GCC-13.2.0/lib -L/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/Mesa/23.1.9-G
CCcore-13.2.0/lib64 -L/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/Mesa/23.1.9-GCCcore-13.2.0/lib -L/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/Boost.MPI/1.83.0-gompi-2023b/lib64 
-L/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/Boost.MPI/1.83.0-gompi-2023b/lib -L/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/SciPy-bundle/2023.11-gfbf-2023b/lib64 -L/cvmfs/softwa
re.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/SciPy-bundle/2023.11-gfbf-2023b/lib -L/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/Python/3.11.5-GCCcore-13.2.0/lib64 -L/cvmfs/software.eessi.io/ver
sions/2023.06/software/linux/x86_64/amd/zen2/software/Python/3.11.5-GCCcore-13.2.0/lib -L/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/FFTW.MPI/3.3.10-gompi-2023b/lib64 -L/cvmfs/software.eessi.io/versions/2023.06/softw
are/linux/x86_64/amd/zen2/software/FFTW.MPI/3.3.10-gompi-2023b/lib -L/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/ScaLAPACK/2.2.0-gompi-2023b-fb/lib64 -L/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/
amd/zen2/software/ScaLAPACK/2.2.0-gompi-2023b-fb/lib -L/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/FlexiBLAS/3.3.1-GCC-13.2.0/lib64 -L/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/
FlexiBLAS/3.3.1-GCC-13.2.0/lib -L/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/GCCcore/13.2.0/lib64 -L/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/GCCcore/13.2.0/lib src/core/reacti
on_methods/tests/CMakeFiles/SingleReaction_test.dir/SingleReaction_test.cpp.o -o src/core/reaction_methods/tests/SingleReaction_test  -Wl,-rpath,/tmp/espresso-bot/easybuild/build/ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75/foss-2023b/easybuild_obj/
src/core:/tmp/espresso-bot/easybuild/build/ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75/foss-2023b/easybuild_obj/src/core/io/writer:/tmp/espresso-bot/easybuild/build/ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75/foss-2023b/easybuild_obj/src/conf
ig:/tmp/espresso-bot/easybuild/build/ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75/foss-2023b/easybuild_obj/src/shapes:/tmp/espresso-bot/easybuild/build/ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75/foss-2023b/easybuild_obj/src/walberla_bridge:/t
mp/espresso-bot/easybuild/build/ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75/foss-2023b/easybuild_obj/_deps/walberla-build/src/field:/tmp/espresso-bot/easybuild/build/ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75/foss-2023b/easybuild_obj/_deps/w
alberla-build/src/vtk:/tmp/espresso-bot/easybuild/build/ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75/foss-2023b/easybuild_obj/_deps/walberla-build/src/blockforest:/tmp/espresso-bot/easybuild/build/ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75/fo
ss-2023b/easybuild_obj/_deps/walberla-build/src/domain_decomposition:/tmp/espresso-bot/easybuild/build/ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75/foss-2023b/easybuild_obj/_deps/walberla-build/src/core  /cvmfs/software.eessi.io/versions/2023.06/sof
tware/linux/x86_64/amd/zen2/software/Boost.MPI/1.83.0-gompi-2023b/lib/libboost_unit_test_framework-mt-x64.so.1.83.0  src/core/espresso_core.so  src/core/io/writer/espresso_hdf5.so  src/config/espresso_config.so  /cvmfs/software.eessi.io/versions/2023.06/
software/linux/x86_64/amd/zen2/software/Boost.MPI/1.83.0-gompi-2023b/lib/libboost_mpi-mt-x64.so.1.83.0  /cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen2/software/Boost.MPI/1.83.0-gompi-2023b/lib/libboost_serialization-mt-x64.so.1.
83.0  -Wl,-rpath-link,/tmp/espresso-bot/easybuild/build/ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75/foss-2023b/easybuild_obj/src/shapes:/tmp/espresso-bot/easybuild/build/ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75/foss-2023b/easybuild_obj/src/walberla_bridge:/tmp/espresso-bot/easybuild/build/ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75/foss-2023b/easybuild_obj/_deps/walberla-build/src/field:/tmp/espresso-bot/easybuild/build/ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75/foss-2023b/easybuild_obj/_deps/walberla-build/src/vtk:/tmp/espresso-bot/easybuild/build/ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75/foss-2023b/easybuild_obj/_deps/walberla-build/src/blockforest:/tmp/espresso-bot/easybuild/build/ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75/foss-2023b/easybuild_obj/_deps/walberla-build/src/domain_decomposition:/tmp/espresso-bot/easybuild/build/ESPResSo/4b9ccc45713137eb16e5f19e5a313b5342080c75/foss-2023b/easybuild_obj/_deps/walberla-build/src/core && :
/cvmfs/software.eessi.io/versions/2023.06/compat/linux/x86_64/usr/bin/ld: warning: libkokkoscontainers.so.4.6, needed by src/core/espresso_core.so, not found (try using -rpath or -rpath-link)
/cvmfs/software.eessi.io/versions/2023.06/compat/linux/x86_64/usr/bin/ld: warning: libkokkoscore.so.4.6, needed by src/core/espresso_core.so, not found (try using -rpath or -rpath-link)
/cvmfs/software.eessi.io/versions/2023.06/compat/linux/x86_64/usr/bin/ld: warning: libkokkossimd.so.4.6, needed by src/core/espresso_core.so, not found (try using -rpath or -rpath-link)
/cvmfs/software.eessi.io/versions/2023.06/compat/linux/x86_64/usr/bin/ld: warning: libheffte.so.2, needed by src/core/espresso_core.so, not found (try using -rpath or -rpath-link)
```

Neves-P commented:
Might be the same seen in https://github.com/easybuilders/easybuild-easyconfigs/pull/19386

jngrad committed:
Remove failing git commands

jngrad commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P commented:
This might have been fixed in https://github.com/easybuilders/easybuild-framework/pull/4435 I'll try using EB >=5.0

jngrad commented:
I have hardcoded `lib64/` in two places in the `.eb` file to circumvent this kind of issue on HPC Vega. I hope you can find a more sustainable way of handling the symlink (or lack thereof) between `lib/` and `lib64/`.

Neves-P committed:
Use EasyBuild 5.1.0

Neves-P committed:
\Merge branch 'easyblock' of https://github.com/jngrad/dev.eessi.io-espresso into easyblock

Neves-P commented:
There is a chance this is already handled, as the symlink is in place since https://github.com/easybuilders/easybuild-framework/pull/4435 in EB 5 implemented symlinks between the two directories. Let's see :crossed_fingers: 


bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P commented:
The emoji killed the smee client on the bot, oops
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P commented:
> I have hardcoded `lib64/` in two places in the `.eb` file to circumvent this kind of issue on HPC Vega. I hope you can find a more sustainable way of handling the symlink (or lack thereof) between `lib/` and `lib64/`.

AFAICS from the logs `CMAKE_INSTALL_LIBDIR` is set to `-DCMAKE_INSTALL_LIBDIR:PATH=lib` but we can try to set `configopts = ' -DCMAKE_INSTALL_LIBDIR:PATH=lib'`

I'll give it a go, I also don't have any better ideas...

Neves-P committed:
Try with

Neves-P commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

jngrad commented:
Locally on HPC Vega, this issue was caused by CMake and Python disagreeing about which lib folder to use. What makes this issue extra annoying is how CMake deploys some `.so` files to two different folders, once under `lib/` and once under `python3.X/site-package/espressomd/`, with different Rpaths. Only the latter are needed. The former are a byproduct that I cannot suppress at the CMake level. When packaging ESPResSo for Fedora, we only deploy the latter. Not sure this information is of any help, but since you're now troubleshooting paths, I might as well inform you of this red herring so you don't waste time investigating the wrong files :sweat_smile:

Neves-P committed:
move 2023.06 subdir to new dev.eessi.io subdir. Comply with EESSI/software-layer-scripts/pull/8

Neves-P commented:
Failed due to EESSI/software-layer-scripts/pull/8

Fixing that now :crossed_fingers: 
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P commented:
> Locally on HPC Vega, this issue was caused by CMake and Python disagreeing about which lib folder to use. What makes this issue extra annoying is how CMake deploys some `.so` files to two different folders, once under `lib/` and once under `python3.X/site-package/espressomd/`, with different Rpaths. Only the latter are needed. The former are a byproduct that I cannot suppress at the CMake level. When packaging ESPResSo for Fedora, we only deploy the latter. Not sure this information is of any help, but since you're now troubleshooting paths, I might as well inform you of this red herring so you don't waste time investigating the wrong files 😅

I see! That is definitely helpful, I'll not pay attention to the stuff in `python3.X/site-package/espressomd/` then :wink: 

I doubt this latest job will succeed given this, but we need to fix the subdirs anyway, so let's see if we see the same error again.

Neves-P committed:
Rename easystacks subdir (temporarily, to confirm effect of change)

Neves-P commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P committed:
Fix silly typo...

Neves-P commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P commented:
I tried running the tests in an interactive build and dropped in the build environment using EasyBuild 5.1.0's resume after error functionality. I can see the "missing" libraries where they should be, but they are not being picked up during the test for some reason. Other libraries are, so I'm not sure what is causing these ones to behave differently.

Need to dig in deeper.

jngrad commented:
Just to clarify, `python3.X/site-package/espressomd/` is the desired output. Anything outside this folder can be safely ignored. We could actually delete them in a post-install step, assuming it's safe to do so (is EB creating man pages or symbolic links that need to be preserved?).

There are two testsuite targets: `check_unit_tests` links the object files directly with the shared libraries, while `check_python` leverages `pypresso` to set the `$PYTHONPATH` such that Python tests find the shared libraries.

Is there a way I can reproduce the build workflow triggered by the bot locally on my machine? The EB files in the first commit of this PR builds, installs and tests ESPResSo without any issue with 4.9.4 and the following commands:
```sh

module load EasyBuild/4.9.4
time eb --force --include-easyblocks ./espresso.py ./ESPResSo-5.0-foss-2023b.eb --robot --prefix="${BUILDDIR}/cache"
module use "${BUILDDIR}/cache/modules/all"
module spider ESPResSo
module load ESPResSo/4b9ccc4-foss-2023b
pypresso -c "import espressomd.version;print(espressomd.version.friendly())"
```
I briefly tried EasyBuild 5.0.x but experienced too many difficulties and eventually went back to 4.9.4.

Neves-P commented:
I'm seeing the same build errors with 4.9.4 when I follow a similar procedure to the bot's workflow (it's a bit harder to fully replicate it, but not impossible)

I've reproduced the error I see in the build workflow by doing the following:

```sh
#setup working directories
mkdir -p eessi_reprod/pr_1
eessi_pr_dir=$HOME/eessi_reprod/pr_1
mkdir -p eessi-manual-builds
eessi_common_dir=${HOME}/eessi_reprod/eessi-manual-builds
cd eessi_reprod/pr_1
git clone https://github.com/jngrad/dev.eessi.io-espresso
git clone https://github.com/eessi/software-layer-scripts
cd dev.eessi.io-espresso
git checkout easyblock
cd ..

#Start eessi container
SINGULARITY_CACHEDIR=${eessi_common_dir}/container_cache software-layer-scripts/eessi_container.sh --access rw -r dev.eessi.io --access ro -r software.eessi.io --nvidia all --host-injections ${eessi_common_dir}/host_injections --save ${eessi_pr_dir}

cd software-layer-scripts

#Source EESSI and setup build environment
export EASYBUILD_ROBOT_PATHS=easyconfigs:
source /cvmfs/software.eessi.io/versions/2023.06/init/bash
module load EESSI-extend

cd ..
# Rename easystack to so EasyBuild/4.9.4 is used
mv dev.eessi.io-espresso/easystacks/software.eessi.io/2023.06/espresso-eb-5.1.0-dev.yml dev.eessi.io-espresso/easystacks/software.eessi.io/2023.06/espresso-eb-4.9.4-dev.yml
cd dev.eessi.io-espresso

# Override EESSI version
module load EasyBuild/4.9.4
# Build from the easystack
eb --easystack easystacks/software.eessi.io/2023.06/espresso-eb-4.9.4-dev.yml --robot --include-easyblocks=easyconfigs/*/*/*.py --software-commit=4b9ccc45713137eb16e5f19e5a313b5342080c75 --debug --search-paths=PATH:easyconfigs/
```

If you give this a try, to you also encounter the error? I have ran this in slurm jobs on the build cluster and also on the RUG HPC and ran into the same errors as in https://github.com/multixscale/dev.eessi.io-espresso/pull/1#issuecomment-2970521369

Indeed I don't think in my attempts the build is landing in `python3.X/site-package/espressomd/`...

_PS: I likely won't be available much today as I'm taking a sick day_

jngrad commented:
Thanks @Neves-P, I'm able to reproduce the issue on HPC Vega with your workflow. Maybe I'll have more time on Friday to continue investigating.

jngrad commented:
Offline discussion with @ocaisa: when linking an executable, on some platforms the linker must recursively check that all symbols in shared objects are defined (even though it's not actually needed, because it's the responsibility of dynamic linking at runtime). Here is the rationale: https://stackoverflow.com/a/35748179. CMake handles that by forwarding private .so file folder paths to consumer executables via the `-Wl,-rpath-link` argument transitively, even if all transitions in a chain are marked as `PRIVATE` dependencies. This was not always the case in older CMake versions, here is the bug report: [cmake/cmake!19556](https://gitlab.kitware.com/cmake/cmake/-/issues/19556). Although this was fixed in CMake 3.16, in some situations the bug can resurface: [cmake/cmake!19707](https://gitlab.kitware.com/cmake/cmake/-/issues/19707). In our case, this affects only the Kokkos shared objects.

Neves-P committed:
Add HeFFTe-2.4.1-foss-2023b.eb

It is a dependency for the builds in #1

jngrad committed:
Use EESSI heFFTe

jngrad committed:
Temporarily revert heFFTe changes

jngrad committed:
Bump ESPResSo version

jngrad committed:
Fix linker errors

jngrad commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

ocaisa commented on easystacks/software.eessi.io/2023.06/espresso-eb-5.1.0-dev.yml:
```diff
@@ -3,4 +3,3 @@ easyconfigs:
       options:
         software-commit: 4b9ccc45713137eb16e5f19e5a313b5342080c75
         include-easyblocks: easyconfigs/*/*/*.py
-  - HeFFTe-2.4.1-foss-2023b.eb
```
This should have been before ESPResSo, not after

jngrad replied:
Noted. Once heFFTe is available, I'll revert the commit that removed the EESSI heFFTe module and shift this line above.

jngrad committed:
Remove auxiliary files

jngrad commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

jngrad committed:
Cleanup espresso.py

jngrad committed:
Bump ESPResSo version

jngrad commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

ocaisa commented on easyconfigs/ESPResSo-foss-2023b-software-commit.eb:
```diff
@@ -0,0 +1,142 @@
+if any(x[0] == 'CUDA' for x in dependencies):
+    cuda_compute_capabilities = ['5.2', '6.0', '7.0', '7.5', '8.0', '8.6', '9.0']
+
+configopts = f' -DCMAKE_SKIP_RPATH=OFF -DESPRESSO_BUILD_TESTS=ON'
```
I'm a bit concerned about `-DCMAKE_SKIP_RPATH=OFF`, is that actually necessary? EESSI (or rather our compiler/linker wrappers) handles the rpathing rather than CMake, so we usually disable all the RPATH behaviour for CMake.

jngrad replied:
That's a leftover from the Fedora specfile I used for guidance. I can remove it.
Should I also patch out the following two R-path CMake options?
https://github.com/espressomd/espresso/blob/8aa60ce/CMakeLists.txt#L798-L808
(they were already present in the 4.2.2 release and weren't an issue)

ocaisa replied:
No, it's fine, I'm probably just afraid of ghosts.

jngrad replied:
For future reference: `-DCMAKE_SKIP_RPATH=OFF` was removed, the two other R-path options in the `CMakeLists.txt` were not removed.

jngrad committed:
More tweaks

jngrad commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

jngrad committed:
Use lib64

jngrad commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

Neves-P commented:
These seem to be test timeouts:
```
The following tests FAILED:
         42 - electrostatic_interactions (Timeout)
         44 - engine_lb (Timeout)
         48 - mass-and-rinertia_per_particle (Timeout)
         68 - rotational-diffusion-aniso (Timeout)
         79 - canonical_ensemble (Timeout)
         81 - lb_pressure_tensor (Timeout)
         82 - ek_fluctuations (Timeout)
         84 - integrator_langevin_stats (Timeout)
         88 - lb_lees_edwards_particle_coupling (Timeout)
         91 - virtual_sites_relative (Timeout)
         93 - virtual_sites_tracers (Timeout)
        105 - lb (Timeout)
        106 - lb_stats (Timeout)
        143 - drude (Timeout)
        150 - lb_circular_couette (Timeout)
        152 - lb_poiseuille_cylinder (Timeout)
        161 - lb_streaming_1_core (Timeout)
        162 - lb_shear (Timeout)
        163 - lb_thermostat (Timeout)
        164 - lb_thermostat_1_core (Timeout)
        165 - lb_momentum_conservation (Timeout)
        166 - lb_momentum_conservation_1_core (Timeout)
        167 - lb_mass_conservation (Timeout)
        169 - p3m_electrostatic_pressure (Timeout)
```

Do these tests need any specific requirements?

jngrad commented:
> These seem to be test timeouts:

Would it be sensible to add `found message matching Timeout` in the failure summary?

These tests are probably timing out because the bot runs in an environment with e.g. 4 cores allocated on a CPU whose spec sheets reports 24 cores. Adding `os.cpu_count()` was probably a bad idea on my part.

jngrad committed:
Revert "Use lib64"

This reverts commit f010aee79cb5b6d8a17f1a1781aecd2c32a98cc3.

Neves-P commented:
Understood! The bot build environment in these jobs is running with 16 cores, perhaps that's not enough? I think `os.cpu_count()` will return 16 in this case, unless somehow this isn't the case inside the build container... I'll check

I think `Timeout` as a keyword can show up during source downloads (which then succeed on a subsequent try from a different URL), but this is something to consider, maybe we can be clever enough in how we search for it and avoid the problem.

jngrad committed:
Factor in oversubscription

jngrad commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

jngrad commented:
Alright, I've changed the oversubscription rule to leave 50% of the cores unallocated during the testsuite, instead of 25%. Each test runs with 2 OpenMP thread per MPI rank, but most of them don't use the second thread, so on average leaving 25% of the core unallocated was sufficient to absorb the extra load from the minority of jobs that actually put a load on the second thread.

We're still actively deploying OpenMP in ESPResSo, and once all algorithms in the software are OpenMP-parallel, we'll adjust the CTest commands to properly factor in the number of OpenMP threads when scheduling the tests.

jngrad commented:
Let me remove the `os.cpu_count()` and `self.cfg.max_parallel` logic completely. Even halving that number doesn't stop timeouts. We'll go back to running the 350 tests one by one, since there is no reliable way of finding a sensible core count to run them in parallel.

jngrad committed:
Disable parallel scheduling of tests

jngrad commented:
bot: build arch:x86_64/amd/zen2 repo:dev.eessi.io

ocaisa commented:
@Neves-P Can we deploy this then? If so I can update the deliverable with a working example

Neves-P commented:
I think so, we can always build the other architectures in a follow up PR

Neves-P commented:
The tarball signing failed, there must be a field that needs to be added to the config. I'll try to solve it

Neves-P commented:
There is something wrong in the signing config that I cannot figure out... I will check and ask for help tomorrow morning.

Neves-P commented:
```sh
 ~ $ ls /cvmfs/dev.eessi.io/espresso/versions/2023.06/software/linux/x86_64/amd/zen2/software/ESPResSo/8aa60cecd56cdd10ab62042c567552f347374f36-foss-2023b/ -l
total 14
dr-xr-xr-x. 2 cvmfs cvmfs-reserved 4096 Jun 24 16:22 bin
dr-xr-sr-x. 3 cvmfs cvmfs-reserved 4096 Jun 24 16:22 easybuild
dr-xr-xr-x. 3 cvmfs cvmfs-reserved 4096 Jun 24 16:22 lib
lrwxrwxrwx. 1 cvmfs cvmfs-reserved    3 Jun 24 16:22 lib64 -> lib
```

The `zen2` build is now available! You'll need to initialize EESSI and do `module use /cvmfs/dev.eessi.io/espresso/versions/2023.06/software/linux/x86_64/amd/zen2/modules/all/` to load the module, but besides that, it should be available wherever `dev.eessi.io` is mounted at :tada: 

I'll merge this PR. A note to ourselves in subsequent PRs:
heFFTe should be available very soon in `software.eessi.io`, so soon we can reapply the change in 69a4b59 and also build for the remaining three microarchitectures.
